### ISSUE TITLE
Fix `types.GetSlotFromHeader` when finalization stalls

### ISSUE THREAD

**Issue by EclesioMeloJunior** [ChainSafe/gossamer#3317]
The following pprof output was extracted when Gossamer was in that region

![image](https://github.com/ChainSafe/gossamer/assets/17255488/85fea9e3-4b9c-4f0b-928e-f82780acdd70)

It is possible to notice that the methods `handleReadyBlock`, `handleWorkersResults`, `processBlockData` and `processBlockDataWithHeaderAndBody` are taking 0s in `flat` and 19.32s in `cum` which indicates that a function they call is hanging 99% of the time and we know that the method `VerifyBlock` is part of the sync flow

If we go deeper in the `processBlockDataWithHeaderAndBody` we can see that the `VerifyBlock` is taking 18.82s 
![image](https://github.com/ChainSafe/gossamer/assets/17255488/fcc57ec9-b8f2-4d22-b4d5-465f23552798)

To be brief the call chain is: 
```

VerifyBlock -> verifyAuthorshipRight -> verifyBlockEquivocation -> b.blockState.GetBlockHashesBySlot -> GetSlotForBlock -> types.GetSlotFromHeader
```

and in the end, the function `GetSlotFromHeader` calls the function `DecodeBabePreDigest` which uses the `scale.Unmarshal` to decode the varying data type and as we can notice it is quite expensive.

![image](https://github.com/ChainSafe/gossamer/assets/17255488/a5181f93-f36c-40cf-a345-553c270f7208)

This process (`DecodeBabePreDigest`) is made by every descendant of the highest finalized block, and we know that at this point 1.41M in the Westend chain the finalization stopped for some period of time.

Basically, each time we want to import a new block we need to verify it, and since we don't have any finalization happening the number of blocks since the latest finalized block only increases and this creates the bottleneck

**Comment by EclesioMeloJunior** [comment]
When the finalization is working the method `VerifyBlocks` takes only 3.94s 
![image](https://github.com/ChainSafe/gossamer/assets/17255488/d11c0188-30b9-4bd1-b213-277c1bb9cdac)

**Comment by timwu20** [comment]
Really good work @EclesioMeloJunior! Showing off those pprof chops 💪
